homeassistant/components/myuplink/sensor.py

```diff
@@ -0,0 +1,160 @@
+"""Sensor for myUplink."""
+from myuplink.models import DevicePoint
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EntityCategory
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import MyUplinkDataCoordinator
+from .const import (
+    DOMAIN,
+    MU_DATAGROUP_DEVICES,
+    MU_DATAGROUP_POINTS,
+    MU_DEVICE_CONNECTIONSTATE,
+    MU_DEVICE_FIRMWARE_CURRENT,
+    MU_DEVICE_FIRMWARE_DESIRED,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up myUplink sensor."""
+
+    entities: list[MyUplinkSensor] = []
+
+    data = hass.data[DOMAIN][config_entry.entry_id]
+    coordinator: MyUplinkDataCoordinator = data["coordinator"]
+
+    # Setup device sensors
+    mu_devices = coordinator.data[MU_DATAGROUP_DEVICES]
+
+    for device_id in mu_devices:
+        entities += [
+            MyUplinkSensor(
+                coordinator=coordinator,
+                data_group=MU_DATAGROUP_DEVICES,
+                data_id=MU_DEVICE_FIRMWARE_CURRENT,
+                device_id=device_id,
+                name="Firmware Current",
+                u_id="firmware_current",
+                diag=True,
+            ),
+            MyUplinkSensor(
+                coordinator=coordinator,
+                data_group=MU_DATAGROUP_DEVICES,
+                data_id=MU_DEVICE_FIRMWARE_DESIRED,
+                device_id=device_id,
+                name="Firmware Desired",
+                u_id="firmware_desired",
+                diag=True,
+            ),
+            MyUplinkSensor(
+                coordinator=coordinator,
+                data_group=MU_DATAGROUP_DEVICES,
+                data_id=MU_DEVICE_CONNECTIONSTATE,
+                device_id=device_id,
+                name="Connection State",
+                u_id="connectionstate",
+                diag=True,
+            ),
+        ]
+
+    # Setup device point sensors
+    mu_device_points = coordinator.data[MU_DATAGROUP_POINTS]
+    for device_id in mu_device_points:
+        for point_id in coordinator.data[MU_DATAGROUP_POINTS][device_id]:
+            point_data: DevicePoint = coordinator.data[MU_DATAGROUP_POINTS][device_id][
+                point_id
+            ]
+            entities.append(
+                MyUplinkSensor(
+                    coordinator=coordinator,
+                    data_group=MU_DATAGROUP_POINTS,
+                    data_id=point_id,
+                    device_id=device_id,
+                    name=point_data.parameter_name,
+                    u_id=point_id,
+                    diag=False,
+                )
+            )
+
+    async_add_entities(entities)
+
+
+class MyUplinkSensor(SensorEntity):
```

Comment:
Inherit from `CoordinatorEntity` too. Then we can simplify the state update here. Look for examples in other integrations that use a `CoordinatorEntity`.

Reply:
Seems to be quite a big refactoring to make that change - i investigated other integrations when i did this setup.

Reply:
Eh, no, it will make things a lot easier in this integration.